### PR TITLE
Fully test, automate and be consistent

### LINKED ISSUE #13: Automate and be consistent
- [ ] Add Travis-CI Automation.
- [ ] Add tests
- [ ] Report Code Coverage to CodeCov, or a similar service.
- [ ] Use the latest ESLint configuration from `teraslice`.

### PR DESCRIPTION
**Pull request by peterdemartini**
Resolves #13

### CODE REVIEW

**Comment by peterdemartini** [comment]
This PR should be ready for review now; I have finished implementing and testing all of the components. Additionally, I tested this version against a running Teraslice.

File: index.js [hunk 1]
```diff
@@ -1,11 +1,21 @@
...
 
-module.exports = function(config) {
-    var teraslice_host = config.host;
+const Jobs = require('./lib/jobs');
+const Job = require('./lib/job');
+const Cluster = require('./lib/cluster');
+const Client = require('./lib/client');
+const Assets = require('./lib/assets');
 
+module.exports = function wrapTerasliceClient(config) {
```
Comment (jsnoble) [inline index.js]: It seems a little dirty to set exports to a function then attach a whole bunch of classes onto the function itself. Yes if you know javascript enough you know that you can do that as technically a function is an object but its confusing to others that don't know the nitty gritty details. You should be returning either an object with all the properties set or return a class that has all the methods formally on it. If you are concerned with backwards compatibility (though its just tjm and teraslice) then the class route would be able to be backwards compatible. However it begs the question of why export the class outside the function like this? The function is supposed to provide configuration for those classes.

File: lib/client.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        return [null, endpoint];
+    }
+    if (rest.length > 2) {
+        return ['Too many arguments passed to client'];
+    }
+    const options = _getRequestOptionsWithData(...rest);
+    return [null, endpoint, options];
+}
+
+function getRequestOptionsWithQuery(endpoint, ...rest) {
```
Comment (jsnoble) [inline lib/client.js]: this is an exact duplicate of the function above with a slight name difference

File: lib/cluster.js [hunk 3]
```diff
@@ -1,50 +1,41 @@
...
+        const validTypes = ['assets', 'slicers', 'ex', 'jobs', 'nodes', 'workers'];
+        const isValid = _.some(validTypes, validType => _.startsWith(type, validType));
+        if (!isValid) {
+            const error = new Error(`"${type}" is not a valid type. Must be one of ${JSON.stringify(validTypes)}`);
+            return Promise.reject(error);
+        }
+        return this.get(`/txt/${type}`, { json: false });
     }
 
-    function post(endpoint, data){
```
Comment (jsnoble) [inline lib/cluster.js:26]: this api is needed. Its used in integration-tests/spec/cases/cluster/api.js and /cases/data/reindex.js. If this functionality is being put somewhere else then we need a pr for teraslice. Also we might need delete soon, but we could add that in later when we really need it

File: lib/job.js [hunk 4]
```diff
@@ -1,167 +1,148 @@
...
-                        }
-                    })
-                    .catch(function(err) {
-                        reject(err);
-                    })
+    waitForStatus(target, intervalMs = 1000, timeoutMs = 0) {
+        // Not all terminal states considered failure.
+        const terminal = {
+            completed: false,
+            failed: true,
```
Comment (jsnoble) [inline lib/job.js]: the list of terminal status are: ['completed', 'stopped', 'rejected', 'failed', 'terminated'] and those should be used as keys in the terminal object. Also completed is a terminal status but its being used to check if its an error scenario so its a bit confusing the way its being called and checked. Maybe change the name of the variable isFatalStatus to better reflect an error based termination vs general job termination

**Review by jsnoble — APPROVED** [review]
tests passed, I also linked this into teraslice for integration tests and those passed as well. LGTM

**Pull request merged**